src/date/DateInput.js

```diff
@@ -34,10 +36,12 @@ const DateInput = createReactClass({
     this.cachedSelectionEnd = this.dateInputInstance.selectionEnd;
     // when popup show, click body will call this, bug!
     const selectedValue = nextProps.selectedValue;
-    this.setState({
-      str: selectedValue && selectedValue.format(nextProps.format) || '',
-      invalid: false,
-    });
+    if (!this.state.hasFocus) {
```

Comment:
This makes `str` & `invalid` un-control. It's better to compare `selectedValue` & `format` changed to setState instead of focus status.

Reply:
I am not sure I understand what you suggest. 

This is done to prevent re-formatting the input while a user is editing the date, i.e. to prevent reformatting `1/12/18` to `01/12/2018` until focus is lost.

As this condition is only required with `formats.length > 2`, I could add this condition. 

Am I missing something here?

Reply:
I mean when props of `selectedValue` or `format` changed. Current code will not makes state update. I suggest to modify like this:
```jsx
if (
  !shallowEqual(this.props.format, format) ||
  (!this.props.selectedValue && selectedValue) ||
  !this.props.selectedValue.isSame(selectedValue, 'date')
) {
  // ...
}
```

Reply:
This won't work because on every key stroke, if the input parses to a valid date, the Calendar component will change the `selectedValue` prop, effectively triggering `componentWillReceiveProps` on DateInput and that will override the date input to a different format while the user is typing.

Reply:
tested. You're correct.